src/main.rs

```diff
@@ -411,74 +210,45 @@ fn main_try() -> Result<(), CargoFlashError> {
-        let target = probe_rs::config::get_target_by_name(chip_name).map_err(|error| {
-            CargoFlashError::ChipNotFound {
-                source: error,
-                name: chip_name.clone(),
-            }
-        })?;
-
-        let loader = build_flashloader(&target, &path)?;
-        (TargetSelector::Specified(target), Some(loader))
-    } else {
-        (TargetSelector::Auto, None)
-    };
-
-    // Try and prepare the probe by opening the probe and selecting the given protocol.
-    let mut probe = open_probe(&opt)?;
-    probe.select_protocol(opt.protocol).map_err(|error| {
-        CargoFlashError::FailedToSelectProtocol {
-            source: error,
-            protocol: opt.protocol,
-        }
-    })?;
+    // Deduce the target to attach to
+    let target_selector = opt.probe_options.get_target_selector()?;
 
-    // Set the SWD or JTAG speed.
-    let protocol_speed = if let Some(speed) = opt.speed {
-        let actual_speed = probe.set_speed(speed).map_err(|error| {
-            CargoFlashError::FailedToSelectProtocolSpeed {
-                source: error,
-                speed,
+    // Attach to specified probe and record metadata
+    let probe = opt.probe_options.attach_to_probe()?;
+    {
+        let protocol_speed = probe.speed_khz();
+        if let Some(speed) = opt.probe_options.speed {
+            if protocol_speed < speed {
+                log::warn!(
+                    "Unable to use specified speed of {} kHz, actual speed used is {} kHz",
+                    speed,
+                    protocol_speed
+                );
             }
-        })?;
-
-        if actual_speed < speed {
-            log::warn!(
-                "Unable to use specified speed of {} kHz, actual speed used is {} kHz",
-                speed,
-                actual_speed
-            );
         }
 
-        actual_speed
-    } else {
-        probe.speed_khz()
-    };
-    // Store the speed in the metadata struct to be able to print it in case of a crash.
-    METADATA.lock().unwrap().speed = Some(format!("{:?}", protocol_speed));
+        // Store probe speed and name in the metadata struct to be able to
+        // print it in case of a crash.
+        METADATA.lock().unwrap().speed = Some(format!("{:?}", protocol_speed));
+        METADATA.lock().unwrap().probe = Some(format!("{:?}", probe.get_name()));
 
-    log::info!("Protocol speed {} kHz", protocol_speed);
+        log::info!("Protocol speed {} kHz", protocol_speed);
+    }
 
-    // Create a new session.
-    // If we wanto attach under reset, we do this with a special function call.
-    // In this case we assume the target to be known.
-    // If we do an attach without a hard reset, we also try to automatically detect the chip at hand to improve the userexperience.
-    let mut session = if opt.connect_under_reset {
+    // Create a new session
+    let mut session = if opt.probe_options.connect_under_reset {
```

Comment:
This logic should be moved to `probe-rs-cli-util` but that would require us to store the `Probe` in `common_options` such that the API is somewhat ergonomic. It is not possible to store non-structopt fields in a `#[structopt] struct { .. }`; i.e. a `Option<Probe>`.

A `lazy_static` maybe?

Reply:
Done via `static mut` for now.